concepts/code-organization/introduction.md

```diff
@@ -0,0 +1,161 @@
+# Introduction
+
+Big projects often span over several files to keep the code organized in logical components.
+
+## Forward Declarations
+C++ code is evaluated procedurally.
+If you want to use a function, it has to be known to the compiler at the moment of usage.
+Sometimes it is not possible to find a linear fashion to lay out the definitions in the source code.
+Take a look at the example below:
+
+```cpp
+int myFunction(int n) {
+    if (n < 10) {
+        return n;
+    } else {
+        return myOtherFunction(n / 10);
+    }
+}
+
+int myOtherFunction(int m) {
+    return myFunction(m / 2);
+}
+```
+
+When `myFunction` is defined, the compiler does not know about `myOtherFunction` yet.
+Unfortunately, the circle-reference problem cannot be solved by switching the order.
+
+C++ offers **forward declarations** to let the compiler know of `myFunction` and `myOtherFunction` before they are defined.
+The compiler assumes that the definition will follow at some later point after the declaration.
+The next example shows how a forward declaration is used for functions.
+
+```cpp
+int myFunction(int n);          // Forward declaration of myFunction
+int myOtherFunction(int m);     // Forward declaration of myOtherFunction
+
+// Definition of myFunction
+int myFunction(int n) {
+    if (n < 0) {
+        return 0;
+    } else {
+        return myOtherFunction(n - 2);
+    }
+}
+
+// Definition of myOtherFunction
+int myOtherFunction(int m) {
+    return myFunction(m / 2);
+}
+```
+
+Attention: the `;` is needed after the declaration, but not necessarily after the definition.
+
+## Header Files
+
+Often, declarations are bundled together in header files. 
+The most common file extension for header files is `.h`.
+Some projects use `.hpp` or skip the extension completely.
```

Comment:
I'm not sure we should go into that much detail. Especially since I aim to keep the C heritage very low-key. The difference between `stdio.h` and `cstdio` would also require some kind of history lesson to be understandable, right?

I would put this concept as the second total the students would see and it is already at  161 lines.

Would this information have been crucial during your first C++ projects?